pkg/controller/validation/ippool_test.go

```diff
@@ -0,0 +1,644 @@
+			subnetInfo: subnetInfo,
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "192.168.1.10", End: "192.168.1.20"},
+			},
+		},
+		{
+			name:       "range start outside subnet",
+			subnetInfo: subnetInfo,
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "10.0.0.1", End: "192.168.1.10"},
+			},
+			expectedErr: "range [10.0.0.1-192.168.1.10] must be a strict subset of the subnet 192.168.1.1/24",
+		},
+		{
+			name:       "range end outside subnet",
+			subnetInfo: subnetInfo,
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "192.168.1.200", End: "192.168.2.10"},
+			},
+			expectedErr: "range [192.168.1.200-192.168.2.10] must be a strict subset of the subnet 192.168.1.1/24",
+		},
+		{
+			name:       "overlapping ranges in same pool",
+			subnetInfo: subnetInfo,
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "192.168.1.10", End: "192.168.1.20"},
+				{Start: "192.168.1.15", End: "192.168.1.25"},
+			},
+			expectedErr: "range [192.168.1.15-192.168.1.25] overlaps with range [192.168.1.10-192.168.1.20]",
+		},
+		{
+			name:       "invalid CIDR range",
+			subnetInfo: subnetInfo,
+			ipRanges: []crdv1beta1.IPRange{
+				{CIDR: "invalid-cidr"},
+			},
+			expectedErr: "invalid cidr invalid-cidr",
+		},
+		{
+			name:       "invalid start IP",
+			subnetInfo: subnetInfo,
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "invalid-ip", End: "192.168.1.10"},
+			},
+			expectedErr: "invalid start ip address invalid-ip",
+		},
+		{
+			name:       "invalid subnet gateway",
+			subnetInfo: &crdv1beta1.SubnetInfo{Gateway: "invalid-gateway", PrefixLength: 24},
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "192.168.1.1", End: "192.168.1.10"},
+			},
+			expectedErr: "invalid gateway address invalid-gateway",
+		},
+		{
+			name:       "invalid IPv4 prefix length",
+			subnetInfo: &crdv1beta1.SubnetInfo{Gateway: "192.168.1.1", PrefixLength: 33},
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "192.168.1.1", End: "192.168.1.10"},
+			},
+			expectedErr: "invalid prefixLength 33",
+		},
+		{
+			name:       "invalid IPv6 prefix length",
+			subnetInfo: &crdv1beta1.SubnetInfo{Gateway: "2001:db8::1", PrefixLength: 129},
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "2001:db8::1", End: "2001:db8::a"},
+			},
+			expectedErr: "invalid prefixLength 129",
+		},
+		{
+			name:       "no subnet info provided",
+			subnetInfo: nil,
+			ipRanges: []crdv1beta1.IPRange{
+				{Start: "192.168.1.1", End: "192.168.1.10"},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := ValidateIPRangesAndSubnetInfo(tt.subnetInfo, tt.ipRanges)
+			if tt.expectedErr != "" {
+				assert.Contains(t, err.Error(), tt.expectedErr)
+			} else {
+				assert.Empty(t, err)
+			}
+		})
+	}
+}
+
+func TestNormalizeRange(t *testing.T) {
+	tests := []struct {
+		name           string
+		ipRange        crdv1beta1.IPRange
+		context        string
+		expectedStart  string
+		expectedEnd    string
+		expectedOrigin string
```

Comment:
I'd suggest having an `expectedRange NormalizedRange` instead and just doing a strict comparison with `assert.Equal`

Reply:
you forgot to remove `expectedStart`, `expectedEnd` and `expectedOrigin` from the struct definition